HootServices: Auth: MapsResource

sumarlidason opened the pull request:
https://github.com/ngageoint/hootenanny/issues/2610

sumarlidason commented:
## Areas of Interest
### Map Access Logic
* GET `@Path("/layers")` [1]
* GET `@Path("/{mapId}/{BBox}")` [2]
* GET `@Path("/{mapId}/mbr")` [2]
* DELETE `@Path("/{mapId}")` [2]
* PUT `@Path("/{mapId}/rename/{name}")` [2]
* PUT `@Path("/{mapId}/move/{folderId : \\d+}")`[2][3]
* GET `@Path("/{mapId}/tags")` [2]

### Folder Access Logic
* GET `@Path("/api/0.6/map/folders")` [4]
* GET `@Path("/linked")` [4]
* POST `@Path("/add/{parentId : \\d+}/{folderName}")` [3]
* DELETE `@Path("/{folderId : \\d+}")` [3]
* PUT `@Path("/{folderId : \\d+}/move/{newParentFolderId : \\d+}")` [3]
* PUT `@Path("/{folderId : \\d+}/rename/{modName}")` [3]

[1] 
Inline SQL
```java
// Owned by the current user
maps.userId.eq(user.getId()).or(
        // or not in a folder // or in a public folder.
        folderMapMappings.id.isNull().or(folderMapMappings.folderId.eq(0L)).or(folders.publicCol.isTrue())
        )
);
```
[2]
Map.java:: isVisibleTo()
```java
return t.get(0, Long.class) == user.getId() || (folderId == null || folderId == 0L) || t.get(2, Boolean.class) == true;
```
[3]
FolderResource::getFolderForUser()
```java
user.getId().longValue() == folder.getUserId().longValue() || folder.getPublicCol().booleanValue()
```
[4]
Inline SQL
```java
folders.userId.eq(user.getId()).or(folders.publicCol.isTrue())
```

brianhatchl commented on hoot-services/src/main/java/hoot/services/controllers/osm/map/FolderResource.java:
```diff
@@ -0,0 +1,369 @@
+        createQuery()
+            .update(folders)
+            .where(folders.parentId.eq(folderId))
+            .set(folders.parentId, folder.getParentId()) // move stuff in this folder to -this folder's- parent.
```
seems like deleting a folder deletes all folders and datasets beneath it in my testing, so this commend confuses me.

sumarlidason replied:
I will add unit testing to confirm logic + bolster confidence

sumarlidason replied:
This does not occur currently, the UI is clearing folders for us w/ successive deletes

brianhatchl commented:
* I created a private folder and then loaded a dataset into it.  I ther refreshed and the dataset was no longer visible, but the private folder was.  The private folder was not visible to another user so appears to be working correctly in that respect.

* What happens if a user creates a private folder under another users public folder and that public folder gets deleted?

* I get some js errors when using context menu actions on root folders, `parentFolder is undefined`

sumarlidason commented on hoot-services/src/main/java/hoot/services/controllers/osm/map/FolderResource.java:
```diff
@@ -0,0 +1,369 @@
+            .execute();
+
+        try {
+            createQuery().insert(folderMapMappings).columns(folderMapMappings.mapId, folderMapMappings.folderId)
```
This function is a bit weird, it contains a **maintenance operation** -- which should probably be converted to foreign key table constraints w/ `ON DELETE CASCADE`.

+ The delete statement removed orphaned folder mapping resources in cases where maps were deleted, but their link was not.
+ The insert statement does the opposite, adding folder links to new maps that haven't been linked to the root folder, `0`. 

From the user's prospective the function is a GET because no changes should be occurring from their request.

Scope Creep Enhancement:
+ Database migration
    + foreign key, ensure integrity of the three way relationship between `maps`, `folders`, and `folder_map_mappings`
+ Refactoring
    + Anywhere new `maps` are created, ensure creation of `folder_map_mappings` with default `folder_id` of `0` to comply with table constraints
    + verify `folder` / `map` removal logic

sumarlidason commented:
>I created a private folder and then loaded a dataset into it. I ther refreshed and the dataset was no longer visible, but the private folder was. The private folder was not visible to another user so appears to be working correctly in that respect.

https://github.com/ngageoint/hootenanny/pull/2682#issuecomment-431169969

> What happens if a user creates a private folder under another users public folder and that public folder gets deleted?

`/folders/add` will prohibit the creation of private folders under public folders, except for root (0)
fixed 554a168

`/folders/add` will prohibit the creation of public folders under private folders
fixed 9194eb1

> I get some js errors when using context menu actions on root folders, parentFolder is undefined

I suspect this was happening because some folders were returning when they shouldn't resulting in broken trees and missing parent folders.
fixed 279006c , 9194eb1

brianhatchl commented:
Is there a way to change a folders visiblity?

brianhatchl commented:
Otherwise looking really :raised_hands: 
One tweak might be to suppress the Public checkbox in the dialog when creating child folders since they will inherit from their parent (and maybe display the folder visibility in a label)

sumarlidason commented:
* changing folder visibility 3235fac22073b794afc483432d1892c6b1ac6213 and https://github.com/ngageoint/hootenanny-ui/commit/46f811fb7d6847ffab4882dc658a21591417709b
<img width="733" alt="image" src="https://user-images.githubusercontent.com/1991770/47331962-ca510a00-d64b-11e8-8749-9081ccdce664.png">

sumarlidason commented:
> One tweak might be to suppress the Public checkbox in the dialog when creating child folders since they will inherit from their parent (and maybe display the folder visibility in a label)

addressed in https://github.com/ngageoint/hootenanny-ui/commit/5b89ffb38cf3cd8cb4af870930515c6b58e2bc45

add folder prompt now more intelligent as of 8cba20df561a2393df9402da0ac95a469c481ab1 ; API also more intelligent / derivative logic

sumarlidason commented:
* add folder to `root`
![screenshot from 2018-10-23 11-26-05](https://user-images.githubusercontent.com/1991770/47372854-a677dd80-d6b8-11e8-9504-60706365c9bb.png)
* , under public
![screenshot from 2018-10-23 11-26-49](https://user-images.githubusercontent.com/1991770/47372888-b5f72680-d6b8-11e8-8fed-be262ea638dd.png)
* , under private
![screenshot from 2018-10-23 11-26-31](https://user-images.githubusercontent.com/1991770/47372904-babbda80-d6b8-11e8-93fb-75766ec53b38.png)
